gh-111140: Improve PyLong_AsNativeBytes API doc example & improve the test (#115380)
python/cpython: Doc/c-api/long.rst

   results if the intent is to match the typical behavior of a C-style
   downcast. No exception is set on truncation.

   Values are always copied as two's-complement, and sufficient buffer will be
   Values are always copied as two's-complement and sufficient buffer will be
   requested to include a sign bit. For example, this may cause an value that
   fits into 8 bytes when treated as unsigned to request 9 bytes, even though
   all eight bytes were copied into the buffer. What has been omitted is the
   zero sign bit, which is redundant when the intention is to treat the value as
   unsigned.
   zero sign bit -- redundant if the caller's intention is to treat the value
   as unsigned.

   Passing zero to *n_bytes* will return the requested buffer size.
   Passing zero to *n_bytes* will return the size of a buffer that would
   be large enough to hold the value. This may be larger than technically
   necessary, but not unreasonably so.

   .. note::

      When the value does not fit in the provided buffer, the requested size
      returned from the function may be larger than necessary. Passing 0 to this

>       function is not an accurate way to determine the bit length of a value.

Reviewer: This (gentle) warning is the main reason for the note. Without it, I think everything here is covered by the preceding two paragraphs.

If you think we don't need to specifically warn people away from calling this instead of `bit_length`, then just remove the whole note.

Author: Good point. The paragraph immediately above the note covers the first sentence anyways. I've replaced the note with the comment about not using it to get a bit length instead.

          // Success!
      }
      else {
          // Overflow occurred, but 'value' contains the truncated
          // lowest bits of pylong.
      }

   The above example may look *similar* to
   :c:func:`PyLong_As* <PyLong_AsSize_t>`
   but instead fills in a specific caller defined type and never raises an
   error about of the :class:`int` *pylong*'s value regardless of *n_bytes*
   or the returned byte count.

   To get at the entire potentially big Python value, this can be used to
   reserve enough space and copy it::

      // Ask how much space we need.
      Py_ssize_t expected = PyLong_AsNativeBits(pylong, NULL, 0, -1);
      if (expected < 0) {
          // A Python exception was set with the reason.
          return NULL;
      }
      assert(expected != 0);  // Impossible per the API definition.
      uint8_t *bignum = malloc(expected);
      if (!bignum) {
          PyErr_SetString(PyExc_MemoryError, "bignum malloc failed.");
          return NULL;
      }
      // Safely get the entire value.
      Py_ssize_t bytes = PyLong_AsNativeBits(pylong, bignum, expected, -1);
      if (bytes < 0 || bytes > expected) {  // Be safe, should not be possible.

>           if (!PyErr_Occurred()) {

Reviewer: ```suggestion
          if (bytes < 0) {
```

One of the few things the C API WG agree on is that we should have consistent enough return values that `PyErr_Occurred` should never be necessary. This is true with this API, so I'd rather not encourage using it.

Your call whether it's better to duplicate the free/return lines in separate blocks.

Author: ironically i had it written that way in my client earlier but shortened it to this for brevity.  agreed, lets not give people the idea that PyErr_Occurred is ever a _good_ idea.

              PyErr_SetString(PyExc_RuntimeError,
                  "Unexpected bignum truncation after a size check.");
          }
          free(bignum);
          return NULL;
      }
      // The expected success given the above pre-check.
      // ... use bignum ...
      free(bignum);

   *endianness* may be passed ``-1`` for the native endian that CPython was
   compiled with, or ``0`` for big endian and ``1`` for little.

>    Returns ``-1`` with an exception raised if *pylong* cannot be interpreted as

Reviewer: Docs team has apparently decided to use `Return` in this context. I much prefer `Returns`, but I'm not part of that group so I don't know the rationale.

Author: Noted. Returns is correct English tense when describing what something does rather than writing code telling a system what to do.  Our docs are not a spec or code, and even a spec would preceed the word return with a qualifier like should or must. :)  I'm going to leave it and let them come through and reword things later for global consistence if they feel otherwise.

   an integer. Otherwise, return the size of the buffer required to store the
   value. If this is equal to or less than *n_bytes*, the entire value was
   copied. ``0`` will never be returned.

   Unless an exception is raised, all *n_bytes* of the buffer will always be
   written. In the case of truncation, as many of the lowest bits of the value
   as could fit are written. This allows the caller to ignore all non-negative
   results if the intent is to match the typical behavior of a C-style
   downcast. No exception is set on truncation.

   Values are always copied as two's-complement and sufficient buffer will be

>    requested to include a sign bit. For example, this may cause an value that

Reviewer: ```suggestion
   requested to include a sign bit. For example, this may cause a value that
```

My original typo, not yours :)

   fits into 8 bytes when treated as unsigned to request 9 bytes, even though
   all eight bytes were copied into the buffer. What has been omitted is the
   zero sign bit -- redundant if the caller's intention is to treat the value
   as unsigned.

>    Passing zero to *n_bytes* will return the size of a buffer that'd at least
>    be large enough to hold the value.

Reviewer: ```suggestion
   Passing zero to *n_bytes* will return the size of a buffer that would
   be large enough to hold the value. This may be larger than technically
   necessary, but not unreasonably so.
```

Two changes here - I think "that'd at least be" is a bit awkward.

The extra line is an alternative to making the `.. note` longer. Originally, I had the note as part of this paragraph, and previous reviewers suggested splitting the warning out so that it would be more obvious. I fear that making the note much longer starts to obscure that.

(I also am trying to avoid saying "might be rounded up to sizeof(size_t)", but perhaps there's a way to phrase that which won't lock us in to particular numbers for all time?)

   .. note::

      When the value does not fit in the provided buffer, the requested size
      returned from the function may be larger than necessary for unsigned use
      cases.  Passing ``n_bytes=0`` to this function is not a way to determine

>       the bit length of a value, the result will merely indicate a buffer size
>       sufficient to hold it.  When passed ``n_bytes=0`` the return value may be
>       larger than technically necessary.

Reviewer: ```suggestion
      the bit length of a value.
```

I think this is best to stop here, and move the extra text into the previous paragraph.
